miden-tx/src/host/tx_authenticator.rs

```diff
@@ -0,0 +1,160 @@
+///
+/// The public keys are defined by [Digest]'s which are the hashes of the actual public keys.
+pub trait TransactionAuthenticator {
+    /// Retrieves a signataure for a specific message as a list of [Felt].
+    /// The request is initiaed by the VM as a consequence of the SigToStack advice
+    /// injector.
+    ///
+    /// - `pub_key`: The public key used for signature generation.
+    /// - `message`: The message to sign, usually a commitment to the transaction data.
+    /// - `account_delta`: An informational parameter describing the changes made to
+    ///   the account up to the point of calling `get_signature()`. This allows the
+    ///   authenticator to review any alterations to the account prior to signing.
+    ///   It should not be directly used in the signature computation.
+    fn get_signature(
+        &self,
+        pub_key: Word,
+        message: Word,
+        account_delta: &AccountDelta,
+    ) -> Result<Vec<Felt>, AuthenticationError>;
+}
+
+// BASIC AUTHENTICATOR
+// ================================================================================================
+
+#[derive(Clone, Debug)]
+pub enum SecretKey {
+    RpoFalcon512(rpo_falcon512::SecretKey),
+}
+
+#[derive(Clone, Debug)]
+/// Represents a signer for [KeySecret] keys
+pub struct BasicAuthenticator<R> {
+    /// pub_key |-> secret_key mapping
+    keys: BTreeMap<Digest, SecretKey>,
+    rng: RefCell<R>,
+}
+
+impl<R: Rng> BasicAuthenticator<R> {
+    #[cfg(feature = "std")]
+    pub fn new(keys: &[(Word, SecretKey)]) -> BasicAuthenticator<rand::rngs::StdRng> {
+        use rand::{rngs::StdRng, SeedableRng};
+
+        let rng = StdRng::from_entropy();
+        BasicAuthenticator::<StdRng>::new_with_rng(keys, rng)
+    }
+
+    pub fn new_with_rng(keys: &[(Word, SecretKey)], rng: R) -> Self {
+        let mut key_map = BTreeMap::new();
+        for (word, secret_key) in keys {
+            key_map.insert(word.into(), secret_key.clone());
+        }
+
+        BasicAuthenticator { keys: key_map, rng: RefCell::new(rng) }
+    }
+}
+
+impl<R: Rng> TransactionAuthenticator for BasicAuthenticator<R> {
+    /// Gets as input a [Word] containing a secret key, and a [Word] representing a message and
+    /// outputs a vector of values to be pushed onto the advice stack.
+    /// The values are the ones required for a Falcon signature verification inside the VM and they are:
+    ///
+    /// 1. The nonce represented as 8 field elements.
+    /// 2. The expanded public key represented as the coefficients of a polynomial of degree < 512.
+    /// 3. The signature represented as the coefficients of a polynomial of degree < 512.
+    /// 4. The product of the above two polynomials in the ring of polynomials with coefficients
+    /// in the Miden field.
+    ///
+    /// # Errors
+    /// Will return an error if either:
+    /// - The secret key is malformed due to either incorrect length or failed decoding.
+    /// - The signature generation failed.
```

Comment:
I would move Falcon-specific comments to the comments of the `get_falcon_signature()` function and here would add a list of supported signature schemes (currently would be just RpoFalcon512).